{% extends 'base.html' %}

{% load static %}
{% load widget_tweaks %}
{% load env %}

{% block breadcrumb %}
	<li class="breadcrumb-item active">Help</li>
{% endblock %}

{% block stylesheet %}
<style>
.vertnav {
	margin: -15px -15px 0px 0px;
}
table td, table td * {
	    vertical-align: top;
}
</style>
{% endblock %}

{% block javascript %}
{% endblock javascript %}

{% block content %}

<ul class="nav flex-column">
	<li class="nav-item vertnav"><a class="nav-link" href="#introduction">Introduction and purpose</a></li>
	<li class="nav-item vertnav"><a class="nav-link" href="#workflow">Workflow</a></li>
	<li class="nav-item vertnav"><a class="nav-link" href="#input">Input & Parameters</a></li>
	<li class="nav-item vertnav"><a class="nav-link" href="#output">Output</a></li>
	<li class="nav-item vertnav"><a class="nav-link" href="#sample">Sample data &amp; sample results</a></li>
	<li class="nav-item vertnav"><a class="nav-link" href="#dbs">Pre-built indices</a></li>
	<li class="nav-item vertnav"><a class="nav-link" href="#browsers">Supported browsers</a></li>
	<li class="nav-item vertnav"><a class="nav-link" href="#citing">Citing PRIMEval</a></li>
	<li class="nav-item vertnav"><a class="nav-link" href="#contact">Contact</a></li>
	<li class="nav-item vertnav"><a class="nav-link" href="#acknowledgments">Acknowledgments</a></li>
</ul>

<h1 id="introduction">Introduction and purpose</h1>

<p style="text-align: justify;"><strong>Multiplex PCR</strong> involves many <strong>challenges</strong> (primer design &amp; validation, concentrations of primers, polymerases, salts (monovalent &amp; divalent) and dNTPS, temperatures, durations and the number of primers and targets in one reaction) and <strong>problems</strong> (formation of unwanted products due to unspecific primer design or targets close to each other, and primer dimer formation by cross-dimerization, self-dimerization or hairpin formation). During <strong>evaluation and validation</strong> of multiplex primer and probe sets, these problems are often observed but not always explainable. However due to the ease of getting whole genome sequences, we have developped <strong>PRIMEval</strong> for an <strong>in silico evaluation of multiplex primer and probe sets</strong>.</p>
<p style="text-align: justify;">PRIMEval <strong>maps primers and probes to whole genomes or contigs</strong> by a background method which can be chosen by the user (Bowtie 1.2.2, BLAST+ 2.7.1 or the Aho–Corasick string-searching algorithm). The user can set a maximum number of <strong>allowed mismatches</strong> in primers and probes and a <strong>maximum amplicon length</strong> generated by two primers. The mapped primers and probes which meet the mismatch criteria are shown as <em>'hits'</em>. If primer hits occur within a certain distance as defined by the maximum amplicon length, a possible amplicon is reported as <em>'result'</em> and probes which bind to an amplicon are also shown. Not all observed or missing products in the wet lab can be explained by this in silico specificity check. Therefore we have implemented a <strong>secondary structure check</strong> using the Primer3 2.3.7 ntthal application which checks primer and probe sets for <strong>cross-dimers, self-dimers and hairpins</strong> depending on the reaction conditions (monovalent &amp; divalent cations, dNTPs, DNA concentration). Problematic oligonucleotides are reported using deltaG and T<sub>M</sub> values and represented in heatmaps.</p>

<p style="text-align: justify;">The <strong>antibiotic resistance gene and virulence factor identifier (ARVI)</strong> is a practical implementation of PRIMEval. ARVI uses a defined set of primers and probes designed against 405 antibiotic resistance genes and 405 virulence factors. Shortly, more than 4000 resistance genes were downloaded from the CARD [1] and ARG-ANNOT [2] databases and virulence factor genes were downloaded from the VFDB [3]. Sequences were clustered with 90% sequence similarity in a one-against-all approach and the representative sequence of each cluster was the one with the most BLAST hits. For both antibiotic resistance genes and virulence factors, the 405 gene clusters with the most hits were used for primer and probe design. Primer and probe design was done using Oli2go in 9 different sets each [1]. These primers and probes are used in PRIMEval for screening purposes. Due to an additional implementation of <strong>mapping files</strong> in PRIMEval, the user can compare the <em>in silico</em> products to the intended products and show additional information (e.g. for antibiotic resistance genes: related to these resistances, resistance mechanism, antibiotic, AMRGeneFamily; for virulence factors: gene related to, from organism). The data was scraped from the CARD or VFDB, respectively.</p>

<p><ol>
	<li>Jia, Baofeng, et al. "CARD 2017: expansion and model-centric curation of the comprehensive antibiotic resistance database." Nucleic acids research (2016): gkw1004.</li>
	<li>Gupta, Sushim Kumar, et al. "ARG-ANNOT, a new bioinformatic tool to discover antibiotic resistance genes in bacterial genomes." Antimicrobial agents and chemotherapy 58.1 (2014): 212-220.</li>
	<li>Chen, Lihong, et al. "VFDB 2016: hierarchical and refined dataset for big data analysis—10 years on." Nucleic acids research 44.D1 (2015): D694-D697.</li>
	<li>Hendling, Michaela, et al. "Oli2go: an automated multiplex oligonucleotide design tool." Nucleic acids research (2018).</li>
</ol></p>

<h1 id="workflow">Workflow</h1>

<p><img src="{% static 'img/workflow.png' %}" alt="PRIMEval workflow diagram" style="display: block; margin-left: auto; margin-right: auto; width: 60%;"></p>

<h1 id="input">Input & Parameters</h1>

<p>Help for all input parameters is shown as popup on the submission page. However, some parameters will be explained here in detail:</p>

<p><ul>
	<li><strong>Method: </strong> Bowtie (default), BLAST or Aho-Corasick. The advantages of each method are described in detail in our publication. Shortly, Aho-Corasick is the recommended &amp; fastest method for a low number of contigs and a low number of mismatches (0 or 1) if no wobbled bases are used in the primers and probes (otherwise a too high number of mismatches might be reported due to the nature of the algorithm). For a higher number of mismatches, Aho-Corasick becomes incredibly slow. Bowtie is the recommended method for any number of mismatches if wobbled primers or probes are used. If screening with a maximum number of 3 mismatches is desired, BLAST performs faster than Bowtie however the number of hits and results is significantly reduced.</li>
	<li><strong>Input files: </strong>Each primer or probe set has to be uploaded as FASTA file with the file ending .fasta. Each primer file should contain all forward and reverse primers from that set and each probe file should contain all probes from that set. The file name corresponds to the set name and the corresponding probe set needs to have the extension _probe.fasta. For example, if the primer set is named abr1.fasta, the corresponding probe set file needs to be named abr1_probes.fasta. The contigs for each strain or sample should be uploaded as one FASTA file with the file ending *.fasta; also the file name represents the sample or strain name (e.g. E_coli_14E.fasta). Mapping files are uploaded as CSV file. While the first 5 columns cannot be renamed and have to stay in the same order (Primer1;Primer2;Probe;Gene;ExpectedSize), an unlimited number of columns can be added to the right. It is best to download our sample data and have a look at the included CSV files.</li>
	<li><strong>Cross-check primer packages: </strong>By default, primers and probes are only matched if the come from the same set. This condition is met if the file names correspond to each other (e.g. package1.fasta and package1_probes.fasta) as described above.</li>
	<li><strong>Check probes only: </strong> This box needs to be checked if no primers are uploaded and only probes are mapped to the uploaded contigs. If primers are uploaded, they will be discarded.</li>
	<li><strong>Secondary structure check: </strong>The secondary structure check performs a cross-dimer, self-dimer and hairpin check on each uploaded oligonucleotide set. For best results, the user needs to adjust parameters for monovalent &amp; divalent cations, dNTPS and DNA concentration.</li>
</ul></p>

<h1 id="output">Output</h1>

<p>Following sections are available in PRIMEval for finished runs:</p>

<p><ul>
	<li><strong>Browse: </strong>A visual representation of forward and reverse primers with the amplicon they form. Probes binding to the amplicon are also represented. This serves as a quick overview for the screening of problematic oligos.</li>
	<li><strong>Results: </strong>All products of forward and reverse primers and eventually matching probes are listed. Mismatches are indicated, and whether primers and probes come from the same or different oligo sets and if more probes potentially bind to the same amplicon.</li>
	<li><strong>Hits: </strong>A list of all primers and probes binding to the uploaded contigs, independently whether they form products in PCR.</li>
	<li><strong>Maps: </strong>(Optional) This section maps the amplicons to the user-defined mapping files.</li>
	<li><strong>Dimer Check: </strong>(Optional) The dimer check shows potentially conflicting oligos by identifying cross-dimers, self-dimers (primers) and hairpins (primers & probes).</li>
	<li><strong>Parameters: </strong>A summary table of anonymized input parameters (e.g. excluding email address).</li>
	<li><strong>Download: </strong>A table containing links to download all output files with (additional) complete information.</li>
</ul></p>

<h1 id="sample">Sample data &amp; sample results</h1>

<p>Here you can find the primer and probe sets as they are used in ARVI (abr and vf) and their mapping files. Additionally, a 45-plex primer and probe set (CARD360) should be used for the secondary structure check. The abr and vf sets should not be used for the secondary structure check because in ARVI, nine 45-plex sets are merged to one 405-plex set.</p>

<p><ul>
	<li><strong>Primer sets: </strong><a href="{% static 'sample_data/abr.fasta' %}" download>abr.fasta</a>, <a href="{% static 'sample_data/vf.fasta' %}" download>vf.fasta</a>, <a href="{% static 'sample_data/CARD360.fasta' %}" download>CARD360.fasta</a></li>
	<li><strong>Probe sets: </strong><a href="{% static 'sample_data/abr_probes.fasta' %}" download>abr_probes.fasta</a>, <a href="{% static 'sample_data/vf_probes.fasta' %}" download>vf_probes.fasta</a>, <a href="{% static 'sample_data/CARD360_probes.fasta' %}" download>CARD360_probes.fasta</a></li>
	<li><strong>Contigs: </strong><a href="{% static 'sample_data/E_coli_15E.fasta' %}" download>E_coli_15E.fasta</a>, <a href="{% static 'sample_data/E_coli_23E.fasta' %}" download>E_coli_23E.fasta</a>, <a href="{% static 'sample_data/E_coli_24E.fasta' %}" download>E_coli_24E.fasta</a>, <a href="{% static 'sample_data/E_coli_26E.fasta' %}" download>E_coli_26E.fasta</a>, <a href="{% static 'sample_data/E_coli_31E.fasta' %}" download>E_coli_31E.fasta</a></li>
	<li><strong>Mapping files: </strong><a href="{% static 'sample_data/abr_map.csv' %}" download>abr_map.csv</a>, <a href="{% static 'sample_data/vf_map.csv' %}" download>vf_map.csv</a></li>
</ul></p>

<p>Here are links to <strong>sample result</strong> pages:</p>

<p><ul>
	<li><a href="{% url 'run' '1280837008' %}">PRIMEval run 1280837008</a>: 18 45-plex primer packages (no probes), all contigs of 9 bacterial whole genome sequences, method: Aho-Corasick, 1 mismatch allowed, max product size: 500 nt, secondary structure check, no cross-check of packages</li>
	<li><a href="{% url 'run' '1565067650' %}">PRIMEval run 1565067650</a>: 1 45-plex primer and probe package, all contigs of 9 bacterial whole genome sequences, method: Bowtie, 1 mismatch allowed in primers and probes, max product size: 1500 nt, secondary structure check</li>
	<li><a href="{% url 'run' '1017748882' %}">ARVI run 1017748882</a>: all contigs of 9 bacterial whole genome sequences, 1 mismatch allowed in primers and 0 mismatches in probes</li>
	<li><a href="{% url 'run' '1603007354' %}">PRIMEval run 1603007354</a>: primer (abr.fasta, vf.fasta) and probe (abr_probes.fasta, vf_probes.fasta) packages and mapping files (abr_map.csv, vf_map.csv) from the sample data section, all contigs of 9 bacterial whole genome sequences, method: BLAST, max product size: 1000, 0 mismatches allowed in primers and probes, no cross-check, no secondary structure check. This example mimicks the behaviour of ARVI with BLAST for oligonucleoide mapping.</li>
</ul></p>

<h1 id="dbs">Databases</h1>

<p>If you are missing your genome of choice (preferentially bigger file sizes), please contact us and we will add it for you.<br />The following pre-built bowtie indices are currently available and were built from the following sources.</p>

<table>
	<tr>
		<th>Species name</th>
		<th>Common name</th>
		<th>Assembly version</th>
		<th>Sequence by</th>
		<th>File size (bytes)</th>
		<th>NCBI Link</th>
		<th>Sequence Link</th>
	</tr>
	<tr>
<td>Ambystoma mexicanum</td>
<td>Axolotl</td>
<td>ASM291563v2</td>
<td>Max Planck Society/University of Kentucky</td>
<td>32810886576</td>
<td><a href="https://www.ncbi.nlm.nih.gov/genome/381">Genome</a></td>
<td><a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/002/915/635/GCA_002915635.2_ASM291563v2/GCA_002915635.2_ASM291563v2_genomic.fna.gz">Sequence</a></td>
</tr>
<tr>
<td>Arabidopsis thaliana</td>
<td>Thale cress</td>
<td>TAIR10.1</td>
<td>The Arabidopsis Information Resource (TAIR)</td>
<td>121164916</td>
<td><a href="https://www.ncbi.nlm.nih.gov/genome/4">Genome</a></td>
<td><a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/735/GCF_000001735.4_TAIR10.1/GCF_000001735.4_TAIR10.1_genomic.fna.gz">Sequence</a></td>
</tr>
<tr>
<td>Caenorhabditis elegans</td>
<td>Caenorhabditis elegans</td>
<td>WBcel235</td>
<td>C. elegans Sequencing Consortium</td>
<td>101540352</td>
<td><a href="https://www.ncbi.nlm.nih.gov/genome/41">Genome</a></td>
<td><a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/002/985/GCF_000002985.6_WBcel235/GCF_000002985.6_WBcel235_genomic.fna.gz">Sequence</a></td>
</tr>
<tr>
<td>Canis lupus familiaris</td>
<td>Domestic dog</td>
<td>CanFam3.1</td>
<td>Dog Genome Sequencing Consortium</td>
<td>2441503953</td>
<td><a href="https://www.ncbi.nlm.nih.gov/genome/85">Genome</a></td>
<td><a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/002/285/GCF_000002285.3_CanFam3.1/GCF_000002285.3_CanFam3.1_genomic.fna.gz">Sequence</a></td>
</tr>
<tr>
<td>Cavia porcellus</td>
<td>Guinea pig</td>
<td>Cavpor3.0</td>
<td>The Genome Sequencing Platform, The Genome Assembly Team</td>
<td>2757684709</td>
<td><a href="https://www.ncbi.nlm.nih.gov/genome/231">Genome</a></td>
<td><a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/151/735/GCF_000151735.1_Cavpor3.0/GCF_000151735.1_Cavpor3.0_genomic.fna.gz">Sequence</a></td>
</tr>
<tr>
<td>Cricetulus griseus</td>
<td>Chinese hamster</td>
<td>CriGri_1.0</td>
<td>Beijing Genomics Institute</td>
<td>2442407538</td>
<td><a href="https://www.ncbi.nlm.nih.gov/genome/2791">Genome</a></td>
<td><a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/223/135/GCF_000223135.1_CriGri_1.0/GCF_000223135.1_CriGri_1.0_genomic.fna.gz">Sequence</a></td>
</tr>
<tr>
<td>Danio rerio</td>
<td>Zebrafish</td>
<td>GRCz11</td>
<td>Genome Reference Consortium</td>
<td>1700419557</td>
<td><a href="https://www.ncbi.nlm.nih.gov/genome/50">Genome</a></td>
<td><a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/002/035/GCF_000002035.6_GRCz11/GCF_000002035.6_GRCz11_genomic.fna.gz">Sequence</a></td>
</tr>
<tr>
<td>Drosophila melanogaster</td>
<td>Common fruit fly</td>
<td>Release 6 plus ISO1 MT</td>
<td>The FlyBase Consortium/Berkeley Drosophila Genome Project/Celera Genomics</td>
<td>145657746</td>
<td><a href="https://www.ncbi.nlm.nih.gov/genome/47">Genome</a></td>
<td><a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/215/GCF_000001215.4_Release_6_plus_ISO1_MT/GCF_000001215.4_Release_6_plus_ISO1_MT_genomic.fna.gz">Sequence</a></td>
</tr>
<tr>
<td>Felis catus</td>
<td>Domestic cat</td>
<td>Felis_catus_9.0</td>
<td>International Cat Genome Sequencing Consortium</td>
<td>2554072303</td>
<td><a href="https://www.ncbi.nlm.nih.gov/genome/78">Genome</a></td>
<td><a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/181/335/GCF_000181335.3_Felis_catus_9.0/GCF_000181335.3_Felis_catus_9.0_genomic.fna.gz">Sequence</a></td>
</tr>
<tr>
<td>Gallus gallus</td>
<td>Chicken</td>
<td>GRCg6a</td>
<td>Genome Reference Consortium</td>
<td>1078736272</td>
<td><a href="https://www.ncbi.nlm.nih.gov/genome/111">Genome</a></td>
<td><a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/002/315/GCF_000002315.6_GRCg6a/GCF_000002315.6_GRCg6a_genomic.fna.gz">Sequence</a></td>
</tr>
<tr>
<td>Heterocephalus glaber</td>
<td>Naked mole-rat</td>
<td>HetGla_female_1.0</td>
<td>Broad Institute</td>
<td>2651538982</td>
<td><a href="https://www.ncbi.nlm.nih.gov/genome/6965">Genome</a></td>
<td><a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/247/695/GCF_000247695.1_HetGla_female_1.0/GCF_000247695.1_HetGla_female_1.0_genomic.fna.gz">Sequence</a></td>
</tr>
<tr>
<td>Homo sapiens</td>
<td>Human</td>
<td>GRCh38.p12</td>
<td>Genome Reference Consortium</td>
<td>3298102737</td>
<td><a href="https://www.ncbi.nlm.nih.gov/genome/51">Genome</a></td>
<td><a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/405/GCF_000001405.38_GRCh38.p12/GCF_000001405.38_GRCh38.p12_genomic.fna.gz">Sequence</a></td>
</tr>
<tr>
<td>Homo sapiens</td>
<td>Human</td>
<td>GRCh38.p13</td>
<td>Genome Reference Consortium</td>
<td>3313087324</td>
<td><a href="https://www.ncbi.nlm.nih.gov/genome/51">Genome</a></td>
<td><a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/405/GCA_000001405.28_GRCh38.p13/GCA_000001405.28_GRCh38.p13_genomic.fna.gz">Sequence</a></td>
</tr>
<tr>
<td>Homo sapiens</td>
<td>Human</td>
<td>GRCh37.p13</td>
<td>Genome Reference Consortium</td>
<td>3275298969</td>
<td><a href="https://www.ncbi.nlm.nih.gov/genome/51">Genome</a></td>
<td><a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/405/GCF_000001405.25_GRCh37.p13/GCF_000001405.25_GRCh37.p13_genomic.fna.gz">Sequence</a></td>
</tr>
<tr>
<td>Macaca mulatta</td>
<td>Rhesus macaque</td>
<td>Mmul_8.0.1</td>
<td>The Genome Institute at Washington University School of Medicine</td>
<td>3313435916</td>
<td><a href="https://www.ncbi.nlm.nih.gov/genome/215">Genome</a></td>
<td><a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/772/875/GCF_000772875.2_Mmul_8.0.1/GCF_000772875.2_Mmul_8.0.1_genomic.fna.gz">Sequence</a></td>
</tr>
<tr>
<td>Mus musculus</td>
<td>House mouse</td>
<td>GRCm38.p6</td>
<td>Genome Reference Consortium</td>
<td>2854241010</td>
<td><a href="https://www.ncbi.nlm.nih.gov/genome/52">Genome</a></td>
<td><a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/635/GCF_000001635.26_GRCm38.p6/GCF_000001635.26_GRCm38.p6_genomic.fna.gz">Sequence</a></td>
</tr>
<tr>
<td>Oryctolagus cuniculus</td>
<td>European rabbit</td>
<td>OryCun2.0</td>
<td>The Genome Sequencing Platform, The Assembly Computation and Development Core Team</td>
<td>2772132550</td>
<td><a href="https://www.ncbi.nlm.nih.gov/genome/316">Genome</a></td>
<td><a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/003/625/GCF_000003625.3_OryCun2.0/GCF_000003625.3_OryCun2.0_genomic.fna.gz">Sequence</a></td>
</tr>
<tr>
<td>Oryza sativa</td>
<td>Asian rice</td>
<td>Build_4.0</td>
<td>International Rice Genome Sequencing Project</td>
<td>387564294</td>
<td><a href="https://www.ncbi.nlm.nih.gov/genome/10">Genome</a></td>
<td><a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/005/425/GCF_000005425.2_Build_4.0/GCF_000005425.2_Build_4.0_genomic.fna.gz">Sequence</a></td>
</tr>
<tr>
<td>Ovis aries</td>
<td>Domestic sheep</td>
<td>Oar_rambouillet_v1.0</td>
<td>Baylor College of Medicine Human Genome Sequencing Center</td>
<td>2906207712</td>
<td><a href="https://www.ncbi.nlm.nih.gov/genome/83">Genome</a></td>
<td><a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/002/742/125/GCF_002742125.1_Oar_rambouillet_v1.0/GCF_002742125.1_Oar_rambouillet_v1.0_genomic.fna.gz">Sequence</a></td>
</tr>
<tr>
<td>Pan troglodytes</td>
<td>Chimpanzee</td>
<td>Clint_PTRv2</td>
<td>The International Chimpanzee Chromosome 22 Consortium</td>
<td>3089233887</td>
<td><a href="https://www.ncbi.nlm.nih.gov/genome/202">Genome</a></td>
<td><a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/002/880/755/GCF_002880755.1_Clint_PTRv2/GCF_002880755.1_Clint_PTRv2_genomic.fna.gz">Sequence</a></td>
</tr>
<tr>
<td>Rattus norvegicus</td>
<td>Common rat</td>
<td>Rnor_6.0</td>
<td>Rat Genome Sequencing Consortium</td>
<td>2906157607</td>
<td><a href="https://www.ncbi.nlm.nih.gov/genome/73">Genome</a></td>
<td><a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/895/GCF_000001895.5_Rnor_6.0/GCF_000001895.5_Rnor_6.0_genomic.fna.gz">Sequence</a></td>
</tr>
<tr>
<td>Saccharomyces cerevisiae S288C</td>
<td>Yeast</td>
<td>R64</td>
<td>Saccharomyces Genome Database</td>
<td>12310392</td>
<td><a href="https://www.ncbi.nlm.nih.gov/genome/15">Genome</a></td>
<td><a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/146/045/GCF_000146045.2_R64/GCF_000146045.2_R64_genomic.fna.gz">Sequence</a></td>
</tr>
<tr>
<td>Sus scrofa</td>
<td>Domestig pig</td>
<td>Sscrofa11.1</td>
<td>The Swine Genome Sequencing Consortium (SGSC)</td>
<td>2533271056</td>
<td><a href="https://www.ncbi.nlm.nih.gov/genome/84">Genome</a></td>
<td><a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/003/025/GCF_000003025.6_Sscrofa11.1/GCF_000003025.6_Sscrofa11.1_genomic.fna.gz">Sequence</a></td>
</tr>
<tr>
<td>Zea mays</td>
<td>Maize</td>
<td>B73_RefGen_v4</td>
<td>maizesequence</td>
<td>2161805065</td>
<td><a href="https://www.ncbi.nlm.nih.gov/genome/12">Genome</a></td>
<td><a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/005/005/GCF_000005005.2_B73_RefGen_v4/GCF_000005005.2_B73_RefGen_v4_genomic.fna.gz">Sequence</a></td>
</tr>
</table>

<p></p>
<p><strong>Following transcriptomes are currently available:</strong></p>
<table>
	<tr>
		<th>Species name</th>
		<th>Common name</th>
		<th>Assembly version</th>
		<th>Sequence by</th>
		<th>File size (bytes)</th>
		<th>NCBI Link</th>
		<th>Sequence Link</th>
	</tr>
<tr>
<td>Arabidopsis thaliana</td>
<td>Thale cress</td>
<td>TAIR10.1</td>
<td>The Arabidopsis Information Resource (TAIR)</td>
<td>95410435</td>
<td><a href="https://www.ncbi.nlm.nih.gov/genome/4">Genome</a></td>
<td><a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/735/GCF_000001735.4_TAIR10.1/GCF_000001735.4_TAIR10.1_rna.fna.gz">Sequence</a></td>
</tr>
<tr>
<td>Caenorhabditis elegans</td>
<td>Caenorhabditis elegans</td>
<td>WBcel235</td>
<td>C. elegans Sequencing Consortium</td>
<td>48002333</td>
<td><a href="https://www.ncbi.nlm.nih.gov/genome/41">Genome</a></td>
<td><a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/002/985/GCF_000002985.6_WBcel235/GCF_000002985.6_WBcel235_rna.fna.gz">Sequence</a></td>
</tr>
<tr>
<td>Cricetulus griseus</td>
<td>Chinese hamster</td>
<td>CriGri_1.0</td>
<td>Beijing Genomics Institute</td>
<td>137509223</td>
<td><a href="https://www.ncbi.nlm.nih.gov/genome/2791">Genome</a></td>
<td><a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/223/135/GCF_000223135.1_CriGri_1.0/GCF_000223135.1_CriGri_1.0_rna.fna.gz">Sequence</a></td>
</tr>
<tr>
<td>Danio rerio</td>
<td>Zebrafish</td>
<td>GRCz11</td>
<td>Genome Reference Consortium</td>
<td>195344836</td>
<td><a href="https://www.ncbi.nlm.nih.gov/genome/50">Genome</a></td>
<td><a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/002/035/GCF_000002035.6_GRCz11/GCF_000002035.6_GRCz11_rna.fna.gz">Sequence</a></td>
</tr>
<tr>
<td>Drosophila melanogaster</td>
<td>Common fruit fly</td>
<td>Release 6 plus ISO1 MT</td>
<td>The FlyBase Consortium/Berkeley Drosophila Genome Project/Celera Genomics</td>
<td>95423852</td>
<td><a href="https://www.ncbi.nlm.nih.gov/genome/47">Genome</a></td>
<td><a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/215/GCF_000001215.4_Release_6_plus_ISO1_MT/GCF_000001215.4_Release_6_plus_ISO1_MT_rna.fna.gz">Sequence</a></td>
</tr>
<tr>
<td>Heterocephalus glaber</td>
<td>Naked mole-rat</td>
<td>HetGla_female_1.0</td>
<td>Broad Institute</td>
<td>267655359</td>
<td><a href="https://www.ncbi.nlm.nih.gov/genome/6965">Genome</a></td>
<td><a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/247/695/GCF_000247695.1_HetGla_female_1.0/GCF_000247695.1_HetGla_female_1.0_rna.fna.gz">Sequence</a></td>
</tr>
<tr>
<td>Homo sapiens</td>
<td>Human</td>
<td>GRCh38.p12</td>
<td>Genome Reference Consortium</td>
<td>589478963</td>
<td><a href="https://www.ncbi.nlm.nih.gov/genome/51">Genome</a></td>
<td><a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/405/GCF_000001405.38_GRCh38.p12/GCF_000001405.38_GRCh38.p12_rna.fna.gz">Sequence</a></td>
</tr>
<tr>
<td>Homo sapiens</td>
<td>Human</td>
<td>GRCh37.p13</td>
<td>Genome Reference Consortium</td>
<td>246455621</td>
<td><a href="https://www.ncbi.nlm.nih.gov/genome/51">Genome</a></td>
<td><a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/405/GCF_000001405.25_GRCh37.p13/GCF_000001405.25_GRCh37.p13_rna.fna.gz">Sequence</a></td>
</tr>
<tr>
<td>Mus musculus</td>
<td>House mouse</td>
<td>458536710</td>
<td>Genome Reference Consortium</td>
<td>2854241010</td>
<td><a href="https://www.ncbi.nlm.nih.gov/genome/52">Genome</a></td>
<td><a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/635/GCF_000001635.26_GRCm38.p6/GCF_000001635.26_GRCm38.p6_rna.fna.gz">Sequence</a></td>
</tr>
<tr>
<td>Saccharomyces cerevisiae S288C</td>
<td>Yeast</td>
<td>R64</td>
<td>Saccharomyces Genome Database</td>
<td>9522619</td>
<td><a href="https://www.ncbi.nlm.nih.gov/genome/15">Genome</a></td>
<td><a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/146/045/GCF_000146045.2_R64/GCF_000146045.2_R64_rna.fna.gz">Sequence</a></td>
</tr>
</table>

<h1 id="browsers">Supported browsers</h1>

<p>We officially support browsers with ECMAScript 6 (ES6) support. We do not recommend accessing PRIMEval via the Tor network due to technical limitations. Following browsers were tested: </p>

<table>
	<tr>
		<th>Browser</th>
		<th>Version</th>
	</tr>
	<tr>
		<td>Google Chrome</td>
		<td>&gt; 62</td>
	</tr>
	<tr>
		<td>Mozilla Firefox</td>
		<td>&gt; 56</td>
	</tr>
	<tr>
		<td>Microsoft Edge</td>
		<td>&gt; 42</td>
	</tr>
	<tr>
		<td>Safari</td>
		<td>&gt; 12</td>
	</tr>
	<tr>
		<td>Opera</td>
		<td>&gt; 50</td>
	</tr>
</table>

<h1 id="citing">Citing PRIMEval</h1>
<p>Please cite: <a href="https://www.nature.com/articles/s41598-019-55883-4">Conzemius, R., Hendling, M., Pabinger, S. and Barišić, I., 2019. PRIMEval: Optimization and screening of multiplex oligonucleotide assays. <i>Scientific Reports</i>, 9(1), pp.1-5.</a></p>

<h1 id="contact">Contact</h1>

{% 'IMPRESSUM_TECH_CONTACT'|env %}

<h1 id="acknowledgments">Acknowledgments</h1>
<p>The PRIMEval software is based on three components: (1) the <strong>PRIMEval client application </strong>(oligonucleotide mapping and secondary structure check tool), (2) the <strong>PRIMEval server application</strong> which handles the communication between the PRIMEval client application and the webserver application, and (3) the <strong>webserver application</strong> which allows public access to PRIMEval.</p>
<p>The <strong>PRIMEval client application</strong> is written in Python 3.4.3. The underlying oligonucleotide mapping algorithms are based on BLAST+ 2.7.1 [1,2], Bowtie 1.2.2 [3] and the Python package pyahocorasick 1.1.8. The secondary structure check uses the ntthal application from Primer3 2.3.7 [4]. Also following third party Python software packages are used: Biopython 1.69 [5], pandas 0.22 [6], NumPy 1.8.2 [7] and XlsxWriter 0.9.3.</p>
<p>The <strong>PRIMEval server application</strong> is written in Python 3.4.3. It uses the Redis 2.8.4 server application and the Python third party software packages redis 2.10.6 and rq 0.12 (Redis Queue). Also following third party Python software packages are used for file and information exchange: urllib3 1.7.1 and PyYAML 3.13.</p>
<p>The <strong>webserver application</strong> is implemented in Python 3.6.7 using the Django 2.1.4 web framework running on an Apache2 webserver and a PostgreSQL database. The client side makes use of following frameworks and libraries: Bootstrap, jQuery, igv.js [8], Font Awesome, DropzoneJS, Popper.js and Google Analytics. The server-side uses following third party Python software packages: matplotlib 3.0.2 [9], seaborn 0.9, NumPy 1.15.4, pandas 0.23.4, urllib3 1.24.1, PyYAML 3.13 and python-magic 0.4.15.</p>

<p><ol>
	<li>Camacho, Christiam, et al. "BLAST+: architecture and applications." BMC bioinformatics 10.1 (2009): 421.</li>
	<li>Altschul, Stephen F., et al. "Basic local alignment search tool." Journal of molecular biology 215.3 (1990): 403-410.</li>
	<li>Langmead, Ben, et al. "Ultrafast and memory-efficient alignment of short DNA sequences to the human genome." Genome biology 10.3 (2009): R25.</li>
	<li>Untergasser, Andreas, et al. "Primer3—new capabilities and interfaces." Nucleic acids research 40.15 (2012): e115-e115.</li>
	<li>Cock, Peter JA, et al. "Biopython: freely available Python tools for computational molecular biology and bioinformatics." Bioinformatics 25.11 (2009): 1422-1423.</li>
	<li>McKinney, Wes. "Data structures for statistical computing in python." Proceedings of the 9th Python in Science Conference. Vol. 445. 2010.</li>
	<li>Oliphant, Travis E. A guide to NumPy. Vol. 1. USA: Trelgol Publishing, 2006.</li>
	<li>Robinson, James T., et al. "Integrative genomics viewer." Nature biotechnology 29.1 (2011): 24.</li>
	<li>Hunter John, D. "Matplotlib: a 2D graphics environment Comput." Sci. Eng 9 (2007): 90-5.</li>
</ol></p>

{% endblock %}
